rfcs/RFC-002-configurable-search-mode.md

```diff
@@ -0,0 +1,157 @@
+| Option 2: Dynamic search mode per query              | 1. Allows queries to be flexible in their desired usage<br/> 2. No need for additional APIs or breaking changes | 1. Adds more complexity for API support<br/> 2. Potentially exposes more challenges when users are debugging queries<br/> 3. The database needs to be customized to store both embeddings and keywords. In certain implementations it could be a memory overhead.<br/> 4. Complicates UX, the user should not have to care  about which search mode they are using (e.g., users don’t configure their search for OpenAI)<br/> 5. Unclear that there are use cases that this is a desirable parameter.<br/> 6. Retaining the VectorIO naming convention is not intuitive for the codebase |
+| Option 3: Separate Provider and API                  | 1. Allows maximum configuration                                                                                 | 1. Larger implementation scope<br/> 2. Requires providers to implement in 3 potential places (inline, remote, and keyword_search_io)<br/> 3. Generalization to hybrid search would logically warrant an additional provider implementation (hybrid_search_io)<br/> 4. Would duplicate a lot of boilerplate code (e.g., configuration of provider) across multiple databases depending on their range of support (listed in reference).                                                                                                                                                   |
+
+## Recommendation and Proposal
+
+Based on our review, the above pros and cons, and how other frameworks approach enabling hybrid and keyword search we
+recommend:
+Option 1: Allow Users to configure the Search Mode through a Provider Config field and an additional `query_string`
+parameter in the API.
+
+### Implementation Detail:
+
+We would extend the RAGQueryConfig in the to accept a `search_mode` parameter:
+
+```python
+@json_schema_type
+class RAGQueryConfig(BaseModel):
+    # This config defines how a query is generated using the messages
+    # for memory bank retrieval.
+    query_generator_config: RAGQueryGeneratorConfig = Field(
+        default=DefaultRAGQueryGeneratorConfig()
+    )
+    max_tokens_in_context: int = 4096
+    max_chunks: int = 5
+    search_mode: str
+```
+
+The Query API is modified to accept `query_string` and `search_mode` parameter within the `EmbeddingIndex` class:
+
+```python
+class EmbeddingIndex(ABC):
+    @abstractmethod
+    async def add_chunks(self, chunks: List[Chunk], embeddings: NDArray):
+        raise NotImplementedError()
+
+    @abstractmethod
+    async def query(
+        self,
+        embedding: NDArray,
+        query_str: Optional[str],
+        k: int,
+        score_threshold: float,
+        search_mode: Optional[str],
```

Comment:
minor nit: why not just `mode`, as the "search" is clear from it's clear from function name (query) ?

Reply:
The PR was updated to use `mode` btw